### ISSUE TITLE
io.BytesIO 'file' (such as with twisted or pytest in py2) -> TypeError: 'unicode' does not have the buffer interface

### ISSUE THREAD

**Issue by altendky** [tqdm/tqdm#589]
The originating issue occurred while [using tqdm in twisted-run tests in py2](https://travis-ci.org/Julian/mkenv/jobs/412040805#L540) (or pytest).  You end up with an `io.BytesIO` as the file which is unable to have a (not-encoded) unicode written to it.

```python
import io
import sys

import tqdm

print(tqdm.__version__, sys.version, sys.platform)

x = [1, 2, 3]

f = io.BytesIO()

for y in tqdm.tqdm(x, file=f):
    pass
```
https://repl.it/@altendky/SnarlingRuralCustomization
```python-traceback
('4.24.0', '2.7.15 (default, Jul 25 2018, 18:24:03) \n[GCC 6.3.0 20170516]', 'linux2')
Traceback (most recent call last):
  File "python", line 12, in <module>
TypeError: 'unicode' does not have the buffer interface
```

I like to document issues first, but I'll go take a look and see if I can straighten out why tqdm is trying to write unicode to an `io.BytesIO` that needs bytes.

**Comment by casperdcl** [comment]
we use `StringIO` for our unit tests

**Comment by Julian** [comment]
Using `StringIO` is incorrect (assuming you mean `io.StringIO` -- `StringIO.StringIO` is a bit of a weird thing which kind of sort of works correctly, but doesn't exist in py3`) -- `sys.stdout.write` takes bytes on py2 -- it will use the default io encoding if you do pass it unicode, doing so though is kind of like relying on automatically coercing unicode to bytes when adding strings together.

If you patch `stdout` with a `StringIO`, *other* code that writes bytes to stdout (correctly) will then blow up, because `io.StringIO` takes only unicode.

**Comment by casperdcl** [comment]
@Julian you can give a `(StringIO|io).StringIO` to `tqdm` explicitly rather than patch `stdout` - which is what we do for all our unit tests

**Comment by altendky** [comment]
@casperdcl, are you suggesting that the proper way to unit test a program that uses tqdm is to provide a special output for tqdm to use during tests?

**Comment by casperdcl** [comment]
"Proper" is ambiguous; it's the approach we've been using (vis. https://raw.githubusercontent.com/tqdm/tqdm/master/tqdm/tests/tests_tqdm.py) -- if there are better alternatives we'd be happy to accept a PR.

**Comment by altendky** [comment]
@Julian, every time I miss the native string thing...  :[  I need to make that stick.  But I'm not clear how to apply that when tqdm is trying to use unicode characters for the progress bar.  Is this one of the cases where you just have to put in an `if PY2: x = x.encode()` before writing it out since you explicitly want encoded unicode (bytes...) in py2?  Though that wouldn't cover being able to pass either a bytes-expecting-file-like or a unicode-expecting-file-like, but perhaps that's a separate issue and there either needs to be both `unicode_file` and `bytes_file` parameters, or just make one mandatory.

@casperdcl, I see for the tests that tqdm doesn't try to use any stdout-capture feature (not sure what nose has for that if anything).  Rather, the `UnicodeIO` or `StringIO` are provided as you mentioned.

https://github.com/tqdm/tqdm/blob/master/tqdm/tests/tests_tqdm.py#L302

**Comment by Julian** [comment]
> Is this one of the cases where you just have to put in an if PY2: x = x.encode() before writing it out since you explicitly want encoded unicode (bytes...) in py2? 

So, IMHO, yes, and that line of code belongs in tqdm (which is why to me, though I haven't followed this super closely) this still strikes me as a tqdm bug, in that it is assuming it can freely write unicode to stdout in Py2 -- because it's doing so, other code is "infected" and forced to have to deal with the potential for unicode to be coming through as well, when yeah, tqdm should be writing bytes to stdout on Py2 (which is the native string "thing" as you put it :) -- code that does I/O should write `str` on both Py2 and Py3).

It sounds like possibly the tqdm maintainers are willing to accept a PR to change that :) -- though the quicker way to unblock ourselves on our end is potentially to also stub out tqdm entirely -- we don't want progress being written out during unit tests, so in theory we should be adding an argument `tqdm`, and setting it to the identity function in our own tests. I suspect that that way is quite a bit quicker (less work) and can be done entirely outside of tqdm.

**Comment by altendky** [comment]
@Julian, if I decide to jump ahead on our tests in mkenv then I'll look at that.  I'm happy to try to improve things here though.  What about my comments on the `file=` parameter tqdm can take?  Should a requirement be made that that file-like can always take native strings?  Or provide a mechanism by which the caller states what type should be used when writing to the file?

**Comment by Julian** [comment]
For the actual file argument personally I'd stick as-is -- so if it wants a single file passed in that it always writes unicode to, going forward that's the only kind of file that should be passed in, and just when it decides "oh I don't have one, let's default to sys.stdout", it should instead be defaulting to `io.TextWrapper(sys.stdout)` internally.

**Comment by Julian** [comment]
So, sorry, even more clearly -- definitely not 2 separate arguments depending on the type of the written data, just one.

**Comment by casperdcl** [comment]
@Julian `tqdm` essentially regularly does `print(u'\u2588', file=sys.stderr)` on both py2 and py3. `u'\u2588'.encode()` fails on py2.

**Comment by casperdcl** [comment]
there's also the `ascii=True` argument which stops `tqdm` making unicode bars.

**Comment by Julian** [comment]
@casperdcl the correct way to write that on Py2 is `print(u'\u2588', file=sys.stderr).encode(whateverEncodingYouWant)` presumably `utf-8`, not without an argument, which will default to `ascii`.

**Comment by altendky** [comment]
@casperdcl, how do you feel about having tqdm write bytes to `sys.stderr` in py2?

I'm writing up a couple tests for that and an adjustment to pass them so we have something concrete to consider.  Though, I'm now hitting the same io operation on a closed file error that I had in mkenv.  We'll see where we end up but it's like tqdm is closing the file it's using for output but when it is `sys.stderr`...  we don't see the message.  The issue only shows up when `io.TextIOWrapper()`ing `sys.stderr`.  Maybe this is only in the tests?

**Comment by altendky** [comment]
Alrighty, there's some actual code to consider.  I still haven't figured out what is actually causing the closure of the files though.  `:[`

**Comment by casperdcl** [comment]
Happy to switch to writing bytes (`.encode("U8")`). @lrq3000 I think implemented most of the current unicode things. `tqdm` never callsv`close()` on its `file`.

**Comment by altendky** [comment]
The `io.TextIOWrapper()` handles the encoding and I try to use the existing encoding specified on the file-like, but presently default to 'utf-8' if that isn't available.  And sure, I couldn't find any code in `tqdm` closing the file...  but something somewhere seems to be doing it.  I put breakpoints in the `close()` method of the wrapper at various times but stack traces were always ambiguous and seemed to be missing frames.

**Comment by casperdcl** [comment]
closed by #598

**Comment by casperdcl** [comment]
@altendky should we also be using e.g. `getattr(sys.stderr, 'buffer', sys.stderr).write`?

**Comment by altendky** [comment]
@casperdcl are you running into some trouble with something?

**Comment by casperdcl** [comment]
@altendky no but just stumbled upon some code which uses this instead of fancy `TextIOWrap`ping. It seems cleaner.

**Comment by altendky** [comment]
I don't think that picking `sys.stdout` vs. `sys.stdout.buffer` would change whether we are passing a `bytes` object vs. `unicode`.  It would also force the thing being substituted in place of the regular `sys.stdout` object to provide a `.buffer` attribute.  The original issue was that we were writing `unicode` to a thing that had been substituted for the 'normal' `sys.stdout` when we should have been writing bytes.
